File: scala/scala-impl/test/org/jetbrains/plugins/scala/lang/lexer/LexerTestBase.java
package org.jetbrains.plugins.scala.lang.lexer;

import com.intellij.lang.LanguageParserDefinitions;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.editor.ex.util.ValidatingLexerWrapper;
import com.intellij.openapi.project.Project;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.plugins.scala.base.ScalaFileSetTestCase;

abstract public class LexerTestBase extends ScalaFileSetTestCase {

    protected LexerTestBase(@NotNull @NonNls String dataPath) {
        super(dataPath);
    }

    @NotNull
    protected Lexer createLexer(@NotNull Project project) {
        return LanguageParserDefinitions.INSTANCE
                .forLanguage(getLanguage())
                .createLexer(project);
    }

    protected void onToken(@NotNull Lexer lexer,
                           @NotNull IElementType tokenType,
                           @NotNull StringBuilder builder) {
        builder.append(tokenType.toString());
        printTokenRange(lexer.getTokenStart(), lexer.getTokenEnd(), builder);
        printTokenText(lexer.getTokenText(), builder);
        builder.append('\n');
    }

    protected void onEof(@NotNull Lexer lexer,
                         @NotNull StringBuilder builder) {
    }

    protected void printTokenRange(int tokenStart, int tokenEnd,
                                   @NotNull StringBuilder builder) {
        builder.append(':').append(' ').append('[')
                .append(tokenStart)
                .append(',').append(' ')
                .append(tokenEnd)
                .append(']').append(',');
    }

    @NotNull
    @Override
    protected String transform(@NotNull String testName,
                               @NotNull String fileText,
                               @NotNull Project project) {
        final Lexer lexer = new ValidatingLexerWrapper(createLexer(project));
        lexer.start(fileText);

        StringBuilder builder = new StringBuilder();

        IElementType tokenType = lexer.getTokenType();
        while (tokenType != null) {
            onToken(lexer, tokenType, builder);

            lexer.advance();
            tokenType = lexer.getTokenType();
        }

        onEof(lexer, builder);
        return builder.toString();
    }

    private static void printTokenText(@NotNull String tokenText,
                                       @NotNull StringBuilder builder) {
        builder.append(' ').append('{')
                .append(tokenText)
                .append('}');
    }
}

File: scala/scala-impl/test/org/jetbrains/plugins/scala/lang/lexer/LexerTestBase.scala
package org.jetbrains.plugins.scala.lang.lexer

import com.intellij.lang.LanguageParserDefinitions
import com.intellij.lexer.Lexer
import com.intellij.psi.tree.IElementType
import org.jetbrains.plugins.scala.base.NoSdkFileSetTestBase

abstract class LexerTestBase extends NoSdkFileSetTestBase {

  protected def createLexer: Lexer =
    LanguageParserDefinitions.INSTANCE.forLanguage(language).createLexer(project)

  protected def onToken(lexer: Lexer, tokenType: IElementType, builder: StringBuilder): Unit = {
    builder.append(tokenType.toString)
    printTokenRange(lexer.getTokenStart, lexer.getTokenEnd, builder)
    printTokenText(lexer.getTokenText, builder)
    builder.append('\n')
  }

  protected def onEof(lexer: Lexer, builder: StringBuilder): Unit = {}

  protected def printTokenRange(tokenStart: Int, tokenEnd: Int, builder: StringBuilder): Unit =
    builder.append(':').append(' ').append('[')
      .append(tokenStart)
      .append(',').append(' ')
      .append(tokenEnd)
      .append(']').append(',')


  override protected def transform(testName: String, fileText: String): String = {
    val lexer = createLexer
    lexer.start(fileText)

    val builder = new StringBuilder()

    var tokenType = lexer.getTokenType
    while (tokenType != null) {
      onToken(lexer, tokenType, builder)

      lexer.advance()
      tokenType = lexer.getTokenType
    }

    onEof(lexer, builder)
    builder.toString
  }

  private def printTokenText(tokenText: String, builder: StringBuilder): Unit =
    builder.append(' ').append('{')
      .append(tokenText)
      .append('}')
}
